demo_nodes_cpp/CMakeLists.txt

```diff
@@ -77,51 +77,149 @@ add_dependencies(services_library)
 add_dependencies(parameters_library)
 add_dependencies(topics_library)
 
-rclcpp_components_register_node(timers_library
-  PLUGIN "demo_nodes_cpp::OneOffTimerNode"
-  EXECUTABLE one_off_timer)
-rclcpp_components_register_node(timers_library
-  PLUGIN "demo_nodes_cpp::ReuseTimerNode"
-  EXECUTABLE reuse_timer)
-
-rclcpp_components_register_node(services_library
-  PLUGIN "demo_nodes_cpp::ServerNode"
-  EXECUTABLE add_two_ints_server)
-rclcpp_components_register_node(services_library
-  PLUGIN "demo_nodes_cpp::ClientNode"
-  EXECUTABLE add_two_ints_client_async)
-
-rclcpp_components_register_node(parameters_library
-  PLUGIN "demo_nodes_cpp::ListParameters"
-  EXECUTABLE list_parameters)
-rclcpp_components_register_node(parameters_library
-  PLUGIN "demo_nodes_cpp::ParameterBlackboard"
-  EXECUTABLE parameter_blackboard)
-rclcpp_components_register_node(parameters_library
-  PLUGIN "demo_nodes_cpp::SetAndGetParameters"
-  EXECUTABLE set_and_get_parameters)
-rclcpp_components_register_node(parameters_library
-  PLUGIN "demo_nodes_cpp::ParameterEventsAsyncNode"
-  EXECUTABLE parameter_events_async)
-rclcpp_components_register_node(parameters_library
-  PLUGIN "demo_nodes_cpp::EvenParameterNode"
-  EXECUTABLE even_parameters_node)
-
-rclcpp_components_register_node(topics_library
-  PLUGIN "demo_nodes_cpp::Talker"
-  EXECUTABLE talker)
-rclcpp_components_register_node(topics_library
-  PLUGIN "demo_nodes_cpp::SerializedMessageTalker"
-  EXECUTABLE talker_serialized_message)
-rclcpp_components_register_node(topics_library
-  PLUGIN "demo_nodes_cpp::Listener"
-  EXECUTABLE listener)
-rclcpp_components_register_node(topics_library
-  PLUGIN "demo_nodes_cpp::SerializedMessageListener"
-  EXECUTABLE listener_serialized_message)
-rclcpp_components_register_node(topics_library
-  PLUGIN "demo_nodes_cpp::ListenerBestEffort"
-  EXECUTABLE listener_best_effort)
+rclcpp_components_register_nodes(
+  PLUGINS
+    "demo_nodes_cpp::ReuseTimerNode"
+    "demo_nodes_cpp::OneOffTimerNode"
+  FROM timers_library
+)
+
+rclcpp_components_add_executable(one_off_timer
+  PLUGINS "demo_nodes_cpp::ReuseTimerNode"
+  FROM timers_library
+)
+
+rclcpp_components_add_executable(reuse_timer
+  PLUGINS "demo_nodes_cpp::ReuseTimerNode"
+  FROM timers_library
+)
+
+install(TARGETS
+  one_off_timer reuse_timer
+  DESTINATION lib/${PROJECT_NAME}
+)
```

Comment:
> I think that this is equally modular and more concise:

And that's completely fair, for sure. I just want to point out that:

1. It obscures an `install` step, implicitly relying on an explicit `install` step for the library it depends on. One may argue registering a component does the same thing, but failure modes are different for a missing entry in ament index and a missing library upon execution.
2. It performs two seemingly unrelated tasks: it registers a component to be discoverable and it creates a standalone executable for that component. One does not need the other. 
3. It bounds us to one component per executable. 

> The name could be better in this case, maybe `rclcpp_components_make_node()` or `rclcpp_components_create_node()`

That's implying there's only one node per executable. Also, it somewhat conflates `executable` and `node` concepts.

------

I tried to picture how it'd be like to actually use this. Every `CMakeLists.txt` from now on will likely `add_library` with components, `ament_target_dependencies` for it and `install` it somewhere. There, what gets done with that library may vary quite a bit.

I can imagine people exposing those components, and there's  `rclcpp_components_register_nodes` or `rclcpp_components_register_components` for that. I can also imagine people not doing so. For instance, one may want to keep tightly coupled components "private" and only provide an executable composing them. Components are the new way of writing ROS2 code, whether it makes sense to dynamically compose them or not. There's `rclcpp_components_add_executable` for that, that purposefully does not register anything.

I can imagine people using their components as standalone executables in the early stages of a project. Specially those that come from a ROS1 setting. I can then picture that same people realizing there's room for system optimization by grouping nodes together. If going full dynamic is an option, then use `launch`, but if not or if it doesn't make sense to go that way `rclcpp_components_target_compose` allows to bundle multiple components in a single executable at build time. I can imagine people building an entire ROS ecosystem in a single process, heavily leveraging intra-process communication.

Reply:
I think it would make sense for it to be `rclcpp_components_register_component` rather than `register_node`, for clarity.

Reply:
> That's implying there's only one component per executable. Also, it somewhat conflates executable and node concepts.

Yeah, but that's the like 99% most common use case in my opinion. I'm not saying we shouldn't build this macro out of more macros that give you more flexibility, but I do think the single component with a single executable for it is the most common case.

I actually don't see a need for multi-component executables, since the users can write their own main to do this, and they can compose the equivalent at runtime with launch or command line tools.

I think it's fine if we have that feature, but it's by no means as important as the single component executable, which I think really every component ought to have, just to preserve the idea that a "composable node" is something that I can either load into a container or run on it's own (where that means it has a corresponding executable).

Reply:
> I think it would make sense for it to be rclcpp_components_register_component rather than register_node, for clarity.

What's the difference in a "component" and a "node"? For me "component" is too generic, it should be "composable_node" or just "node".

Reply:
> I do think the single component with a single executable for it is the most common case.

It is.

> I actually don't see a need for multi-component executables, since the users can write their own main to do this

True, but it's going to be boilerplate most of the time -- unless one gets very specific about NodeOptions.

Reply:
Well, not necessarily, they can have any number of additional arguments to their constructor, for example, maybe you have a scenario where you want two nodes to have access to the same instance of a library (maybe one that reads data from a camera or something. You couldn't specify that with dynamic composition, but you could if you were manually composing. Also you might want to get fancy with how many and what kinds of executors you want to use.

Reply:
> > I actually don't see a need for multi-component executables, since the users can write their own main to do this
>
> True, but it's going to be boilerplate most of the time -- unless one gets very specific about NodeOptions.

I think this type of "boilerplate" won't show up too much if we expect this to be a rare use-case.

Seems like we all agree that it's good to have tools in place to support this feature, but we should have the simpler, recommended, API built on top. Now we just need to determine the color of the shed.